File: ps/3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int i, left, right, closest = 100000000;
        sort(nums.begin(), nums.end());
        for(i=0;i<nums.size()-1;i++) {
            int local_closest;
            left = i+1;
            right = nums.size()-1;
            while (left < right) {
                int cur = nums[i] + nums[left] + nums[right] - target;
                if (abs(cur) < abs(closest)) {
                    closest = cur;
                }
                if (cur < 0) left++;
                else if (cur > 0) right--;
                else return target;
            }
        }
        return closest+target;
    }
};

File: ps/3sum-closest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        min_val = sys.maxsize
        for idx, val in enumerate(nums):
            i = idx + 1
            j = len(nums) - 1
            while i < j:
                candid = (nums[i] + nums[j] + val)
                diff = target - candid
                if abs(diff) < abs(min_val):
                    min_val = diff
                if diff > 0:
                    i += 1
                elif diff < 0:
                    j -= 1
                else:
                    return target

        return target - min_val
        
